diff --git a/src/daemon/goadaemon.c b/src/daemon/goadaemon.c
index 0eb372e..db770b4 100644
--- a/src/daemon/goadaemon.c
+++ b/src/daemon/goadaemon.c
@@ -535,14 +535,14 @@ add_config_file (GoaDaemon     *self,
                   if (provider_type != NULL)
                     provider = goa_provider_get_for_provider_type (provider_type);
 
+                  needs_update = g_key_file_remove_group (key_file, groups[n], NULL) || needs_update;
+
                   if (provider == NULL)
                     {
-                      g_warning ("Unsupported account type %s for ID %s (no provider)", provider_type, id);
+                      g_debug ("Unsupported account type %s for ID %s (no provider)", provider_type, id);
                       goto cleanup_and_continue;
                     }
 
-                  needs_update = g_key_file_remove_group (key_file, groups[n], NULL);
-
                   error = NULL;
                   if (!goa_utils_delete_credentials_for_id_sync (provider, id, NULL, &error))
                     {
@@ -562,7 +562,7 @@ add_config_file (GoaDaemon     *self,
             }
           else
             {
-              needs_update = g_key_file_remove_key (key_file, groups[n], "SessionId", NULL);
+              needs_update = g_key_file_remove_key (key_file, groups[n], "SessionId", NULL) || needs_update;
             }
 
           g_hash_table_insert (group_name_to_key_file_data,
@@ -647,7 +647,7 @@ update_account_object (GoaDaemon           *self,
   provider = goa_provider_get_for_provider_type (type);
   if (provider == NULL)
     {
-      g_warning ("Unsupported account type %s for identity %s (no provider)", type, identity);
+      g_debug ("Unsupported account type %s for identity %s (no provider)", type, identity);
       goto out;
     }
 
